In [11]:
import numpy as np
import random
import time

In [2]:
def add_obstacle(grid, start, end):
    x0, y0 = start
    x1, y1 = end

    if x0 == x1:  # Vertical obstacle
        for y in range(min(y0, y1), max(y0, y1) + 1):
            grid[x0, y] = 1
    elif y0 == y1:  # Horizontal obstacle
        for x in range(min(x0, x1), max(x0, x1) + 1):
            grid[x, y0] = 1

    return grid

def has_at_least_one_zero(grid):
    
    for row in grid:
        if np.all(row == 1):
            return False
            
    for col in grid.T:
        if np.all(col == 1):
            return False
    return True

In [3]:
grid = np.zeros((31, 31), dtype='int16')
num_obstacles = round(0.09 * len(grid.flatten()))
for _ in range(num_obstacles):
    while True:
        start = (random.randint(1, grid.shape[0]-2), random.randint(1, grid.shape[0]-2))
        end = (random.randint(1, grid.shape[0]-2), random.randint(1, grid.shape[0]-2))
        if start != end:
            if start[0]==end[0] or start[0]==end[1]:
                temp_grid = grid.copy()  
                temp_grid = add_obstacle(temp_grid, start, end)
                if has_at_least_one_zero(temp_grid):
                    grid = temp_grid
                    break

In [4]:
print(grid)

[[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0]
 [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 1 1 1 1 1 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 0 0]
 [0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0]
 [0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0]
 [0 0 0 0 0 0 1 0 1 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 0 0 0 0]
 [0 0 0 0 0 0 1 0 1 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 1 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 1 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 0 0]
 [0 0 0 0 0 0 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 0 0]
 [0 0 0 0 0 0 1 0 1 1 1 1

In [5]:
import numpy as np
import random

def create_maze(grid_size):    
    grid = np.ones((grid_size, grid_size), dtype=int)

    start_x, start_y = (1, 1)
    # grid[start_x, start_y] = 2

    directions = [(0, 2), (2, 0), (0, -2), (-2, 0)]
    
    def carve(x, y):
        
        random.shuffle(directions)
        for dx, dy in directions:
            nx, ny = x + dx, y + dy
            if 0 < nx < grid_size and 0 < ny < grid_size and grid[nx, ny] == 1:

                grid[nx, ny] = 0
                grid[x + dx // 2, y + dy // 2] = 0
                carve(nx, ny)

    carve(start_x, start_y)
    
    # grid[:, 0] = 0
    # grid[0, :] = 0
   
    return grid

In [30]:
grid_size = 16
    
maze_grid = create_maze(grid_size)

for row in maze_grid:
    print(' '.join(['*' if cell == 1 else '$' if cell == 2 else ' ' for cell in row]))

* * * * * * * * * * * * * * * *
*       *           *          
*   *   * * *   *   *   *   * *
*   *           *   *   *      
* * * * * * * * *   *   * * *  
*       *       *   *   *      
*   *   * * *   *   * * *   *  
*   *   *       *       *   *  
*   *   *   *   * * *   *   *  
*   *       *   *       *   *  
*   * * * * * * *   * * *   * *
*               *   *          
*   * * * * *   *   * * * * *  
*           *                  
* * * * *   * * * * * * * * * *
*                              


In [14]:
grid_size = 100
    
maze_grid2 = create_maze(grid_size)

for row in maze_grid2:
    print(' '.join(['*' if cell == 1 else '$' if cell == 2 else ' ' for cell in row]))

* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
*       *           *       *                       *       *               *               *       *           *                   *       *           *               *                           *  
*   *   *   * * *   *   *   *   * * * * * * *   *   *   * * *   * * * * *   *   * * * * *   *   *   *   *   *   * * * * * * * * *   *   * * *   * * *   *   * * *   *   *   * * * * * * * * *   *   *  
*       *       *   *   *   *           *   *   *       *       *           *       *   *       *       *   *                       *       *       *   *   *   *   *   *           *       *   *      
*   * * * * * * *   *   *   * * * * *   *   *   * * * * *   * * *   * * * * * * *   *   * * * * * * * * *   * * * * * * * * * * *   * * *   * * *   *   *   *   *   * * * * * * *   *   *   *   * * *  


In [15]:
class Maze_MDP:
    def __init__(self, maze, start, goal):
        self.maze = maze  # 0 for open space, 1 for walls
        self.start = start
        self.goal = goal
        self.rows = maze.shape[0]
        self.cols = maze.shape[1]
        
        # Define actions: up, down, left, right
        self.actions = {
            0: (-1, 0),  # up since indices are smaller when we go up
            1: (1, 0),   # down
            2: (0, -1),  # left
            3: (0, 1)    # right
        }
        
        self.reward_goal = 77
        self.reward_move = -1
        self.reward_obstacle = -17

    def is_valid_state(self, state):
        i, j = state
        return 0 <= i < self.rows and 0 <= j < self.cols and self.maze[i, j] == 0

    def transition(self, state, action):
        dx, dy = self.actions[action]
        new_state = (state[0] + dx, state[1] + dy)
        if self.is_valid_state(new_state):
            return new_state
        return state  # stay in the same state if invalid

    def get_reward(self, state):
        if state == self.goal:
            return self.reward_goal
        if self.maze[state[0], state[1]] == 1:
            return self.reward_obstacle  # only applicable if we move to an obstacle
        return self.reward_move

    def print_policy(self, policy):
        actions = ['U', 'D', 'L', 'R']
        for i in range(self.rows):
            for j in range(self.cols):
                if self.maze[i, j] == 1:  # Wall
                    print(' ', end=" | ")
                elif (i, j) == self.goal:
                    print('G', end=" | ")  # Goal
                else:
                    print(actions[policy[i, j]], end=" | ")
            print()
            print((j+1)*"__|_")

In [32]:
mdp = Maze_MDP(maze_grid, start=(1, 1), goal=(maze_grid.shape[0]-1, maze_grid.shape[1]-1))

In [25]:
mdp2 = Maze_MDP(maze_grid2, start=(1, 1), goal=(maze_grid2.shape[0]-1, maze_grid2.shape[1]-1))

## Value Iteration

In [26]:
def value_iteration(mdp, gamma=0.9, theta=1e-6):
    V = np.zeros((mdp.rows, mdp.cols))
    V[mdp.goal] = mdp.reward_goal  # Set value for the goal state    
    
    while True:    
        delta = 0
        for i in range(mdp.rows):
            for j in range(mdp.cols):
                if mdp.maze[i, j] == 1:
                    # V[i, j] = mdp.reward_obstacle
                    continue
                
                if (i, j) == mdp.goal:
                    continue

                v = V[i, j]
                V[i, j] = max(
                    mdp.get_reward((i, j)) + gamma * V[mdp.transition((i, j), a)]
                    for a in range(len(mdp.actions))
                )
                delta = max(delta, abs(v - V[i, j]))                
                
        if delta < theta:
            break
    return V


def extract_policy(mdp, V):
    gamma = 0.9
    policy = np.zeros((mdp.rows, mdp.cols), dtype=int)  # Initialize policy
    for i in range(mdp.rows):
        for j in range(mdp.cols):
            if (i, j) == mdp.goal:
                continue
            # Find the action that maximizes the value
            
            best_action = max(range(len(mdp.actions)), 
                              key=lambda a: mdp.get_reward((i, j)) + 
                                            gamma * V[mdp.transition((i, j), a)])            
            policy[i, j] = best_action  # Assign the best action to the policy
    return policy

# for 16x16

In [33]:
start_time = time.time()
V = value_iteration(mdp)
policy_v = extract_policy(mdp, V)
print("Time Taken : ",time.time()-start_time, " seconds")
mdp.print_policy(policy_v)

Time Taken :  0.06136488914489746  seconds
  |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   | 
__|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|_
  | R | R | D |   | R | R | R | R | D |   | R | R | D | L | L | 
__|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|_
  | U |   | D |   |   |   | U |   | D |   | U |   | D |   |   | 
__|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|_
  | U |   | R | R | R | R | U |   | D |   | U |   | R | R | D | 
__|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|_
  |   |   |   |   |   |   |   |   | D |   | U |   |   |   | D | 
__|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|_
  | D | L | L |   | R | R | D |   | D |   | U |   | D | L | L | 
__|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|_
  | D |   | U |   |   |   | D |   | D |   |   |   | D |   | U | 
__|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|_
  | D |   | U |   | D | L | L |   | R | R | D |

# for 100x100

In [29]:
start_time = time.time()
V = value_iteration(mdp2)
policy_v = extract_policy(mdp2, V)
print("Time Taken : ",time.time()-start_time, " seconds")
mdp2.print_policy(policy_v)

Time Taken :  4.549448490142822  seconds
  |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   | 
__|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|_
  | U | U | U |   | U | U | U | U | U |   | U | U | U |   | U | U | U | U | U | U | U | U | U | U | U |   | U | U | U |   | U | U | U | U | U | U | U |   | U

## Policy Iteration

In [34]:
def random_policy(mdp):    
    policy = np.random.choice([0, 1, 2, 3], size=(mdp.rows, mdp.cols))
    return policy

def evaluate_policy(mdp, policy, gamma=0.9, threshold=1e-4):    
    value = np.zeros((mdp.rows, mdp.cols))
    
    while True:
        delta = 0
        new_value = np.copy(value)
        
        for i in range(mdp.rows):
            for j in range(mdp.cols):
                state = (i, j)
                if not mdp.is_valid_state(state):
                    continue
                
                action = policy[i, j]
                next_state = mdp.transition(state, action)
                reward = mdp.get_reward(next_state)
                
                # Use the Bellman equation to update the value function
                new_value[i, j] = reward + gamma * value[next_state[0], next_state[1]]
                delta = max(delta, abs(new_value[i, j] - value[i, j]))
        
        value = new_value
        if delta < threshold:
            break
    
    return value

def improve_policy(mdp, value, gamma=0.9):    
    policy = np.zeros((mdp.rows, mdp.cols), dtype=int)
    
    for i in range(mdp.rows):
        for j in range(mdp.cols):
            state = (i, j)
            if not mdp.is_valid_state(state):
                continue
            
            state_values = []
            for action in mdp.actions:
                next_state = mdp.transition(state, action)
                reward = mdp.get_reward(next_state)
                state_value = reward + gamma * value[next_state[0], next_state[1]]
                state_values.append(state_value)
            
            # Choose the action that maximizes the value
            best_action = np.argmax(state_values)
            policy[i, j] = best_action
    
    return policy

def policy_iteration(mdp, gamma=0.9, threshold=1e-4):    
    policy = random_policy(mdp)
    while True:
        value = evaluate_policy(mdp, policy, gamma, threshold)  # Policy evaluation
        new_policy = improve_policy(mdp, value, gamma)  # Policy improvement
        
        # If the policy stops changing, we've converged
        if np.array_equal(policy, new_policy):
            break
        
        policy = new_policy
    
    return value, policy        

# for 16x16

In [35]:
start_time = time.time()
value, policy_p = policy_iteration(mdp)
print("Time Taken : ",time.time()-start_time, " seconds")
mdp.print_policy(policy_p)

Time Taken :  2.6886188983917236  seconds
  |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   | 
__|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|_
  | R | R | D |   | R | R | R | R | D |   | R | R | D | L | L | 
__|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|_
  | U |   | D |   |   |   | U |   | D |   | U |   | D |   |   | 
__|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|_
  | U |   | R | R | R | R | U |   | D |   | U |   | R | R | D | 
__|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|_
  |   |   |   |   |   |   |   |   | D |   | U |   |   |   | D | 
__|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|_
  | D | L | L |   | R | R | D |   | D |   | U |   | D | L | L | 
__|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|_
  | D |   | U |   |   |   | D |   | D |   |   |   | D |   | U | 
__|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|_
  | D |   | U |   | D | L | L |   | R | R | D | 

# for 100x100

In [36]:
start_time = time.time()
value, policy_p = policy_iteration(mdp2)
print("Time Taken : ",time.time()-start_time, " seconds")
mdp2.print_policy(policy_p)

Time Taken :  206.6202392578125  seconds
  |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   | 
__|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|_
  | U | U | U |   | U | U | U | U | U |   | U | U | U |   | U | U | U | U | U | U | U | U | U | U | U |   | U | U | U |   | U | U | U | U | U | U | U |   | U

# Benckmarking

## Comparison between policy iteration and value iteration

### For 16x16

- **Policy Iteration** : 2.686 seconds **Value Iteration** :  0.061 seconds

### For 100x100

- **Policy Iteration** : 206.620 seconds **Value Iteration** :  4.549 seconds

#### -> Value Iteration is faster than Policy Iteration